tests/udp/udp-netapi-tests.c

```diff
@@ -0,0 +1,295 @@
+    ng_udp_hdr_t *udp_hdr;
+    uint8_t *payload_data;
+
+    ipv6 = build_ipv6_recv_header(src, dst, data_len);
+
+    if (ipv6 == NULL) {
+        return NULL;
+    }
+
+    payload = ng_pktbuf_add(ipv6, NULL, sizeof(ng_udp_hdr_t) + data_len,
+                            NG_NETTYPE_UDP);
+
+    if (payload == NULL) {
+        /* no need to release, we reset packet buffer on fail anyways */
+        return NULL;
+    }
+
+    udp_hdr = payload->data;
+    payload_data = ((uint8_t *)payload->data) + sizeof(ng_udp_hdr_t);
+    udp_hdr->src_port = byteorder_htons(src_port);
+    udp_hdr->dst_port = byteorder_htons(dst_port);
+    udp_hdr->checksum = byteorder_htons(checksum);
+    udp_hdr->length = byteorder_htons(sizeof(ng_udp_hdr_t) + data_len);
+
+    memcpy(payload_data, data, data_len);
+
+    return payload;
+}
+
+static ng_pktsnip_t *build_udp_recv_exp_packet(const ng_ipv6_addr_t *src,
+        uint16_t src_port, const ng_ipv6_addr_t *dst, uint16_t dst_port,
+        uint16_t checksum, char *data, size_t data_len)
+{
+    ng_pktsnip_t *ipv6, *udp, *payload;
+    ng_udp_hdr_t *udp_hdr;
+
+    ipv6 = build_ipv6_recv_header(src, dst, data_len);
+
+    if (ipv6 == NULL) {
+        return NULL;
+    }
+
+    udp = ng_udp_hdr_build(ipv6, (uint8_t *)&src_port, sizeof(src_port),
+                           (uint8_t *)&dst_port, sizeof(dst_port));
+
+    if (udp == NULL) {
+        /* no need to release, we reset packet buffer on fail anyways */
+        return NULL;
+    }
+
+    udp_hdr = udp->data;
+    udp_hdr->checksum = byteorder_htons(checksum);
+    udp_hdr->length = byteorder_htons(sizeof(ng_udp_hdr_t) + data_len);
+
+    payload = ng_pktbuf_add(udp, data, data_len, NG_NETTYPE_UNDEF);
+
+    return payload;
+}
+
+static void test_send_ipv6(void)
+{
+    TEST_ASSERT(_udp_pid != KERNEL_PID_UNDEF);
+
+    const ng_ipv6_addr_t src = TEST_SRC_IPV6;
+    const ng_ipv6_addr_t dst = TEST_DST_IPV6;
+    ng_pktsnip_t *in = build_udp_send_packet(&src, TEST_SRC_PORT,
+                                             &dst, TEST_DST_PORT,
+                                             TEST_DATA, sizeof(TEST_DATA));
+    kernel_pid_t exp_senders[] = { _udp_pid };
+    ng_pktsnip_t *exp_out[] = { build_udp_send_packet(&src, TEST_SRC_PORT,
+                                                      &dst, TEST_DST_PORT,
+                                                      TEST_DATA, sizeof(TEST_DATA)) };
+
+    TEST_ASSERT_NOT_NULL(exp_out[0]);
+
+    ng_udp_hdr_t *exp_udp = exp_out[0]->next->data;
+
+    exp_udp->length = byteorder_htons(sizeof(ng_udp_hdr_t) + sizeof(TEST_DATA));
+
+    TEST_ASSERT_NOT_NULL(in);
+    TEST_ASSERT_EQUAL_INT(NG_NETTEST_SUCCESS,
+                          ng_nettest_send(_udp_pid, in, 1, exp_senders,
+                                          (const ng_pktsnip_t **)exp_out,
+                                          NG_NETTYPE_IPV6, NG_NETREG_DEMUX_CTX_ALL));
+    ng_pktbuf_release(exp_out[0]);
+    TEST_ASSERT(ng_pktbuf_is_empty());
```

Comment:
That's seems a bunch of asserts in one test case. It is adviseable trying to test one aspect in each test case.
This influences error masking and backtracking.